utils/data.hpp

```diff
@@ -0,0 +1,143 @@
+  Qt::NoPen, Qt::SolidLine, Qt::DashLine, Qt::DotLine,
+    Qt::DashDotLine, Qt::DashDotDotLine, Qt::CustomDashLine
+};
+const QStringList PEN_CAP_STR = {
+    "SquareCap", "FlatCap", "RoundCap"
+};
+const QList<Qt::PenCapStyle> PEN_CAP_VAL = {
+  Qt::SquareCap, Qt::FlatCap, Qt::RoundCap
+};
+const QStringList PEN_JOIN_STR = {
+  "BevelJoin", "MiterJoin", "RoundJoin"
+};
+const QList<Qt::PenJoinStyle> PEN_JOIN_VAL = {
+  Qt::BevelJoin, Qt::MiterJoin, Qt::RoundJoin
+};
+const QStringList BRUSH_STYLE_STR = {
+  "SolidPattern", "Dense1Pattern", "Dense2Pattern",
+    "Dense3Pattern", "Dense4Pattern", "Dense5Pattern",
+    "Dense6Pattern", "Dense7Pattern", "NoBrush",
+    "HorPattern", "VerPattern", "CrossPattern",
+    "BDiagPattern", "FDiagPattern", "DiagCrossPattern",
+    "LinearGradientPattern", "RadialGradientPattern",
+    "ConicalGradientPattern"
+};
+const QList<Qt::BrushStyle> BRUSH_STYLE_VAL = {
+  Qt::SolidPattern, Qt::Dense1Pattern, Qt::Dense2Pattern,
+    Qt::Dense3Pattern, Qt::Dense4Pattern, Qt::Dense5Pattern,
+    Qt::Dense6Pattern, Qt::Dense7Pattern, Qt::NoBrush,
+    Qt::HorPattern, Qt::VerPattern, Qt::CrossPattern,
+    Qt::BDiagPattern, Qt::FDiagPattern, Qt::DiagCrossPattern,
+    Qt::LinearGradientPattern, Qt::RadialGradientPattern,
+    Qt::ConicalGradientPattern
+};
+const QStringList TEXT_ALIGN_STR = {
+    "AlignLeft", "AlignRight", "AlignTop",
+    "AlignBottom", "AlignCenter"
+};
+const QList<Qt::AlignmentFlag> TEXT_ALIGN_VAL = {
+  Qt::AlignLeft, Qt::AlignRight, Qt::AlignTop,
+    Qt::AlignBottom, Qt::AlignCenter
+};
+const QStringList TEXT_STYLE_STR = {
+  "StyleNormal", "StyleItalic", "StyleOblique"
+};
+const QList<QFont::Style> TEXT_STYLE_VAL = {
+  QFont::StyleNormal, QFont::StyleItalic, QFont::StyleOblique
+};
+const QStringList TEXT_WEIGHT_STR = {
+  "Thin", "ExtraLight", "Light", "Normal", "Medium",
+    "DemiBold", "Bold", "ExtraBold", "Black"
+};
+const QList<QFont::Weight> TEXT_WEIGHT_VAL = {
+  QFont::Thin, QFont::ExtraLight, QFont::Light, QFont::Normal, QFont::Medium,
+    QFont::DemiBold, QFont::Bold, QFont::ExtraBold, QFont::Black
+};
+
+class Data
+{
+private:
+    Canvas* mainCanvas; // Pointer to the main canvas of the program
+    QString currentDirectory;   // Current file loaded into the project
+
+    void GetTo(QTextStream&, QChar);  // Helper function moves the given text stream up to the given character
+    QPoint* GetPoints(QList<qint32>&);    // Helper function converts a list of integers to a pointer to a block of qpoitns
+
+    void ExtractFont(QTextStream&, QFont&);  // Extract data on the shape's font from the text stream specified
+
+    Shape* ConstructLine(QString&, QList<qint32>&/*, QPen*/);  // Construct a line-type using the given data
+    Shape* ConstructShape(QString&, QList<qint32>&/*, QPen, QBrush*/); // Construct a 2-d type shape with the given data
+    Shape* ConstructText(QString&, QList<qint32>&, QString&, QFont&); // Construct a text-type shape with the given data
```

Comment:
These 3 funcs should return more specific types.

Reply:
The problem with that though is that the methods don't construct just one datatype.  Like, in `ConstructLine` we have:

```suggestion
    if(type == LINE)
    {
       return new Line(QPoint(dimensions[0], dimensions[1]), QPoint(dimensions[2], dimensions[3]));
    }
    else if(type == POLYLINE)
    {
        return new Polyline(getPoints(dimensions), uint32_t(dimensions.size() / 2));
    }
```
which means we might return a `Line` or a `Polyline`, so it has to return the base class pointer.

Reply:
Sorry I think that suggestion box was formatted strangly, but hopefully you get the picture ;)